packages/yew/src/dom_bundle/mod.rs

```diff
@@ -5,147 +5,72 @@
 //! In order to efficiently implement updates, and diffing, additional information has to be
 //! kept around. This information is carried in the bundle.
 
-mod app_handle;
 mod bcomp;
 mod blist;
 mod bnode;
 mod bportal;
 mod bsuspense;
 mod btag;
 mod btext;
+mod traits;
+mod utils;
 
-#[cfg(test)]
-mod tests;
+use gloo::utils::document;
+use web_sys::{Element, Node};
 
-use self::bcomp::BComp;
-use self::blist::BList;
-use self::bnode::BNode;
-use self::bportal::BPortal;
-use self::bsuspense::BSuspense;
-use self::btag::BTag;
-use self::btext::BText;
+use crate::html::AnyScope;
+use crate::html::NodeRef;
+use crate::virtual_dom::VNode;
 
-pub(crate) use self::bcomp::{ComponentRenderState, Mountable, PropsWrapper, Scoped};
+use bcomp::BComp;
+use blist::BList;
+use bnode::BNode;
+use bportal::BPortal;
+use bsuspense::BSuspense;
+use btag::BTag;
+use btext::BText;
+use traits::{DomBundle, Reconcilable};
+use utils::{insert_node, test_log};
 
-#[doc(hidden)] // Publically exported from crate::app_handle
-pub use self::app_handle::AppHandle;
 #[doc(hidden)] // Publically exported from crate::events
 pub use self::btag::set_event_bubbling;
-#[cfg(test)]
-#[doc(hidden)] // Publically exported from crate::tests
-pub use self::tests::layout_tests;
 
-use crate::html::AnyScope;
-use crate::NodeRef;
-use web_sys::{Element, Node};
-
-trait DomBundle {
-    /// Remove self from parent.
-    ///
-    /// Parent to detach is `true` if the parent element will also be detached.
-    fn detach(self, parent: &Element, parent_to_detach: bool);
-
-    /// Move elements from one parent to another parent.
-    /// This is for example used by `VSuspense` to preserve component state without detaching
-    /// (which destroys component state).
-    fn shift(&self, next_parent: &Element, next_sibling: NodeRef);
-}
+/// A Bundle.
+///
+/// Each component holds a bundle that represents a realised layout, designated by a VNode.
+///
+/// This is not to be confused with [BComp], which represents a component in the position of a
+/// bundle layout.
+#[derive(Debug)]
+pub(crate) struct Bundle(BNode);
```

Comment:
`Bundle` seems to be ambiguous with `DomBundle`?
However, I cannot come up with a better name here.

Reply:
Is there a reason not to use `BNode` directly? None of this is public API either way. Also, could the placeholder just be an empty `BNode::List(BList::new())`? Inserting empty text nodes as placeholder messes with `MutationObserver` - though getting rid of it in other places would be a separate issue.

Reply:
If the benefit is to have the traits be less visible and only expose `Bundle` from `dom_bundle`, then I actually don't oppose just calling it `Bundle` and renaming the `DomBundle` trait to something more meaningful.

Reply:
If `BNode` is used directly, then both `Reconcilable` and `DomBundle` needs to be `pub(crate)` which I am fine both ways, but I guess the original visibility is limited to the module, then it would be better to keep it.

> DomBundle -> Stateful

Name `Stateful` is already used by CompStateInner.

Reply:
I have picked the name `ReconcileTarget` to replace `DomBundle`.

Reply:
> be an empty `BNode::List(BList::new())`?

I think the original decision to always insert a text node on first render is to maintain a valid next_sibling NodeRef in initial rendering and removing it may cause unintended side effects.

I have tested with this by replacing it with `VNode::List(BList::new())` and the test wouldn't pass.